aten/src/TH/THRandom.c

```diff
@@ -305,6 +305,51 @@ double THRandom_standard_gamma(THGenerator *_generator, double alpha) {
   }
 }
 
+// This is identical to THRandom_standard_gamma but also stores the
+// reparameterized gradient wrt alpha in grad_alpha.
+double THRandom_standard_gamma_with_grad(THGenerator *_generator, double alpha,
+                                         double *grad_alpha) {
+  double scale = 1.0;
+  double scale_grad = 0.0;
+
+  // Boost alpha for higher acceptance probability.
+  if(alpha < 1.0) {
+    const double u = 1 - uniform_double(_generator);
+    scale = pow(u, 1.0 / alpha);
+    scale_grad = -scale * log(u) / (alpha * alpha);
+    alpha += 1.0;
+  }
+
+  // This implements the acceptance-rejection method of Marsaglia and Tsang (2000)
+  // doi:10.1145/358407.358414
+  const double d = alpha - 1.0 / 3.0;
+  const double d_grad = 1.0;
+  const double c = 1.0 / sqrt(9.0 * d);
+  const double c_grad = -0.5 * c / d;
+  for(;;) {
+    double x, y;
+    do {
+      x = THRandom_normal(_generator, 0.0, 1.0);
+      y = 1.0 + c * x;
+    } while(y <= 0);
+    const double v = y * y * y;
+    const double u = 1 - uniform_double(_generator);
+    const double e = 1.0 - v + log(v);
+    const double accept = 0.5 * x * x + d * e;
+    if(log(u) < accept) {
+      const double y_grad = c_grad * x;
+      const double v_grad = 3 * y * y * y_grad;
+      const double e_grad = (1 / v - 1) * v_grad;
+      const double accept_grad = d_grad * e + d * e_grad;
+      const double dv = d * v;
+      const double dv_grad = d_grad * v + d * v_grad      // Pathwise part.
+                           + (dv - alpha) * accept_grad;  // Acceptance part.
```

Comment:
This is the trickiest part. This roughly follows [Naesseth et al. (2017)](http://proceedings.mlr.press/v54/naesseth17a/naesseth17a.pdf) but computes an exact reparameterized gradient rather than approximating. Naesseth et al. seem to be missing the analytical baseline `alpha` in `(dv - alpha)` which is easy to compute. Note that `accept = log(acceptance ratio)`, so we're using the log trick in multiplying by `accept_grad` (and avoiding an expensive `exp()`).